### ISSUE TITLE
Allow users to change the vert.x context propagation through threads

### ISSUE THREAD

**Issue by jvican** [eclipse-vertx/vert.x#3619]
#### Describe the feature

The `VertxImpl` implementation uses a `ThreadLocal` called `stickyContext` to propagate the context when `getOrCreateContext` or `getContext` are called outside of a pool of vert.x threads.

However, there are other ways that the state can be propagated in other executor pools. For example, effect libraries such as [Monix](https://github.com/monix/monix) have their own way of propagating state local to an execution with `TaskLocal` -- this is an equivalent mechanism to the context locals that vert.x uses internally to work around the limitation that handlers can be run in different handlers and yet be part of the same semantic execution trace.

I'd like to request that vert.x allows users to bring their own mechanism to propagate the vert.x context.

That can be achieved in two ways:

- Making `getContext` part of the `Vertx` API just like `getOrCreateContext` is -- this way, I can create a `WrappedVertx` extending `Vertx`, wrapping the default `vertx: VertxImpl` and overriding `getContext` and `getOrCreateContext`.
- Adding a new abstraction `LocalManager`, similar in name and spirit to `ClusterManager`, that I can implement as a service and that vertx loads through SPI. The default instance of `LocalManager` will use the thread local sticky context implementation with the weak reference.

I prefer the second solution over the first one. Do you have any thoughts?

#### Contribution

I can open a pull request implementing this feature.

**Comment by vietj** [comment]
what are the actual use cases for this feature ?

**Comment by vietj** [comment]
stickyContext is not a way to propagate a context, it is a way to guarantee that the same context will always be used when called from a non vertx thread in order to avoid races.

**Comment by jvican** [comment]
> stickyContext is not a way to propagate a context, it is a way to guarantee that the same context will always be used when called from a non vertx thread in order to avoid races.

Yeah, that's propagation for me. To be more explicit, Vert.x wants to make sure it propagates the context for all runnables that run in a thread not managed by the framework.

> what are the actual use cases for this feature ?

I'm building a framework using Monix, the mentioned effect system library, that propagates local state with a mechanism different than thread locals.
